packages/contractkit/src/utils/web3-utils.ts

```diff
@@ -12,24 +12,95 @@ import { CeloProvider } from './tx-signing'
 
 const debug = debugFactory('kit:web3:utils')
 
+export function generateAccountAddressFromPrivateKey(privateKey: string): string {
+  if (!privateKey.toLowerCase().startsWith('0x')) {
+    privateKey = '0x' + privateKey
+  }
+  return new Web3().eth.accounts.privateKeyToAccount(privateKey).address
+}
+
 // This web3 has signing ability.
-export async function addLocalAccount(web3: Web3, privateKey: string): Promise<Web3> {
+export function addLocalAccount(web3: Web3, privateKey: string): Web3 {
   const celoProvider = new CeloProvider(privateKey)
+  const existingProvider = web3.currentProvider
+  let providerEngine: Web3ProviderEngine
+  if (existingProvider instanceof Web3ProviderEngine) {
+    debug('Existing provider is already a Web3ProviderEngine')
+    providerEngine = addLocalAccountToExistingProvider(
+      existingProvider as Web3ProviderEngine,
+      celoProvider
+    )
+  } else {
+    providerEngine = createNewProviderWithLocalAccount(existingProvider, celoProvider)
+    web3.setProvider(providerEngine)
+  }
+
+  providerEngine.start()
+  debug('Providers configured')
+  return web3
+}
+
+function addLocalAccountToExistingProvider(
```

Comment:
I think we can do this whole design better...

We can have:
`web3.currentProvider instanceOf CeloProvider`

And then create a CeloProvider, that internally proxies everything to the ProviderEngine, has a reference to the Transport/original provider. Knows the accounts, has a `addAccount()` method... etc.

Reply:
the current code seems unnecessarily complex

Reply:
Good point. I am changing `CeloProvider` to `CeloPrivateKeyWalletProvider` and adding a new `CeloProvider` which hides all this complexity.